### PR TITLE
Implements irreducible characters of (full) Permutation Group

### PR DESCRIPTION
**Pull request by kalmarek**
function `character(lambda::Partition, p::perm) computes the value of `lambda`-th character of `S_n` on `p`, assuming that `lambda` is partition of `n` and `p` belongs to `PermutationGroup(N)`. 

Notes:

* Strictly necessary for this functionality are `Partition`, `IntPartitions`, `partitionseq` and `MN1inner` (Murnaghan-Nakayama formula), i.e. all up to https://github.com/Nemocas/Nemo.jl/compare/master...kalmarek:enh/SNreps?expand=1#diff-2eeee0faa8e88f21d4b924e361206827R241
* `YoungTableau(lambda::Partition)` and `SkewDiagram(lam::Partition, mu::Partition)` are just nice-to-haves at this point.
* ~`noPartitions(n)` (counting number of partitions of n) is not type stable:~
  - ~for `n<395` will return `Int`,~
  - ~above that `BigInt`~
  - ~as is will be correct only when `Int = Int64`.~
~Stabler, but slower solution could be to return always `BigInt`~
* functions `emb, emb!` embedding naturally permutations `S_n→S_{n+k}` are also something not strictly necessary, but required for e.g. `YoungSymmetrisers` which are not ready at this point
* values of characters are cached in a dictionary, which means if You call `character` with all checks turned off You may store completely wrong value, which may or may not lead to completely meaningless computations until You restart julia
* Please comment!

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
For noPartitions, we already have numpart().

**Comment by kalmarek** [comment]
Thanks, fixed!

**Comment by thofma** [comment]
I'll have a look during the next days.

**Comment by kalmarek** [comment]
no need for haste

**Comment by thofma** [comment]
Can you add tests for the character values? They are also incorrect for large values. You need to make the values of type `fmpz`, since they quickly overflow.
```
julia> p = Partition(collect(10:-1:1));

julia> q = PermutationGroup(55)();

julia> character(p, q)
-6310668977852907520
```
The correct result is `44261486084874072183645699204710400`

**Comment by kalmarek** [comment]
I haven't used `fmpz` yet, so I'm not sure if this "lazy" solution is the performant one;

Also after doing some computations I think a much better solution to evaluation would be something along
```julia
type Character <: Function
    p::Partition
end
(chi::Character)(g::GroupElem) = MN1Inner(chi.p, permtype(g), 1)
```
because it allows easily for e.g.
```julia
function (chi::Character)(r::GroupRingElem)
    RG = parent(r)
    d = findn(r.coeffs)
    return sum(chi.(RG.basis[d]) .* r.coeffs[d])
end
```

File: src/generic/PermGroups.jl [hunk 1]
```diff
@@ -388,6 +394,14 @@ doc"""
 order(a::perm) = lcm([length(c) for c in cycles(a)])
 
 doc"""
+    permtype(a::perm, rev=true)
+> Returns the type of permutation `a`, i.e. lengths of disjoint cycles building
+> `a`. This fully determines conjugacy class of `a`. The lengths are sorted in
```
Comment (thofma) [inline src/generic/PermGroups.jl]: the conjugacy class

File: src/generic/YoungTabs.jl [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+# Implemented following RuleAsc (Algorithm 3.1) from
+#    "Generating All Partitions: A Comparison Of Two Encodings"
+# by Jerome Kelleher and Barry O’Sullivan, ArXiv:0909.2331
+
+doc"""
+   IntPartitions(n::Int)
+> Returns an iterator over all integer `Partition`s of `n`. They come in
+> ascending order. See also `Combinatorics.partitions(n)`.
+"""
+immutable IntPartitions
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: What does `IntPartitions` stand for?

File: src/generic/YoungTabs.jl [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+##############################################################################
+#
+#   Partition sequences and Murnaghan-Nakayama formula
+#
+##############################################################################
+
+doc"""
+    partitionseq(lambda::Partition)
+> Returns a sequence (as `BitVector`) of `false`s and `trues`s constructed from
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: `true`s

File: src/generic/YoungTabs.jl [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+> `charvals::Dict{Tuple{BitVector,Vector{Int}}, Int}`.
+> This is an implementation (with slight modifications) of the
+> Murnaghan-Nakayama formula as described in
+>
+>     Dan Bernstein,
+>     "The computational complexity of rules for the character table of Sn"
+>     _Journal of Symbolic Computation_, 37(6), 2004, p. 727-748.
+"""
+function MN1inner(R::BitVector, mu::Partition, t::Int,
+        charvals=Dict{Tuple{BitVector,Vector{Int}}, fmpz}())
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: indentation

File: src/generic/YoungTabs.jl [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+function haslneigh(xi::SkewDiagram, i::Int, j::Int)
+   if j == 1
+      return false
+   else
+      return inskewdiag(xi, i, j) && inskewdiag(xi, i, j-1)
+   end
+end
+
+doc"""
+    hasdneigh(xi::SkewDiagram, i::Int, j::Int)
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: Why `hasdneigh` and not `hasbneigh`?

File: src/generic/YoungTabs.jl [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+doc"""
+    isrimhook(xi::SkewDiagram)
+> Checks if `xi` represents a rim-hook diagram, i.e. its diagram is
+> edge-connected and contains no $2\times 2$ squares.
+"""
+function isrimhook(xi::SkewDiagram)
+   i = 1
+   j = xi.lam[1]
+   while i ≠ length(xi.lam) && j ≠ 1
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: No non-ASCII characters please.

File: src/generic/YoungTabs.jl (hunk 2 above)
Comment (kalmarek) [inline src/generic/YoungTabs.jl]: Integer partitions; Is simple `Partitions` enough?

Comment (thofma) [inline src/generic/YoungTabs.jl]: Yeah, or `PartitionsIter` for partitions iterator. Or just `Partitions`.

**Comment by kalmarek** [comment]
As a matter of fact I had those verbose names in place before, but wanted to "be more in line" with practice in julia ;-) 

Thanks for the review.

**Comment by kalmarek** [comment]
Two more things (I'm fine with leaving both as is now)
* What do You think about the `Character` `type`/`immutable` approach to characters? Do You know anyone more experienced with computations on characters who could chime in and give some suggestions? 
* Do we accept non-ASCII characters in the output of e.g. show?
I found using 
```julia
function subscriptify(n::Int)
    subscript_0 = Int(0x2080) # Char(0x2080) -> subscript 0
    return join([Char(subscript_0 + i) for i in reverse(digits(n))])
end
function Base.show(io::IO, p::Partition)
    uniq = unique(p.part)
    mults = [count(i -> i == u, p.part) for u in uniq]
    str = join((string(u)*subscriptify(m) for (u,m) in zip(uniq, mults)))
    print(io, str)
end
```
very convenient with e.g. `3₂2₁1₂` representing `Partition([3, 3, 2, 1, 1])`.

**Comment by thofma** [comment]
I think non-ASCII output is fine. As long as I don't have to type it.

The character type approach is a good idea. We do the same for morphism types in Hecke. By the way, you don't have to make it a subtype of `Function`. Once a type is callable you can also use the dot syntax:
```julia
julia> ZZ.([1,2,3])
3-element Array{Nemo.fmpz,1}:
 1
 2
 3
```
One more thing: I think for partitions of `n` you should save the number `n` itself in the type, so you don't have to compute `sum()` all the time.

**Comment by kalmarek** [comment]
ok, I'll add `n` field to `Partition`. I think we can then get rid of `n` in `YoungTableau`, as this will be accessible through `Y.part.n`.

After few tries (and fails) `Character <: AbstractCharacter` doesn't seem to cut it.
More likely we would have to go for `PermCharacter`, `DifferentGroupCharacter` etc, as all of these will require different fields to store the information necessary for evaluation.
This kind of defeats my aim (to easily evaluate on different types) as we can not add methods to `AbstractCharacter`. As for now I use (which I don't particularly like)
```julia
for T in [PermCharacter, DirectProdCharacter]
   @eval begin
      function (chi::$T){S}(X::GroupRingElem{S})
         RG = parent(X)
         z = zero(S)
         result = z
         for i in 1:length(X.coeffs)
            if X.coeffs[i] != z
               result += chi(RG.basis[i])*X.coeffs[i]
            end
         end
         return result
      end
   end
end
```

**Comment by thofma** [comment]
Yes, we also ran into this problem with overloading abstract polynomial types. Your workaround should be fine.

**Comment by thofma** [comment]
After talking to Bill and Fredrik today, we think non-ASCII output is okay, but there should be a way to get ASCII output. Maybe using a global flag like `non_ascii_printing`. But we can add this later. It does not have to be done in this PR here.

**Comment by thofma** [comment]
If there are no objections, I will merge this this evening.

**Comment by wbhart** [comment]
No objections.

**Comment by wbhart** [comment]
I get the following error.

youngtabs.skewdiags...Error During Test
  Test threw an exception of type UndefVarError
  Expression: Nemo.matrix_repr(Partition([1],false) /
Partition(Int[],false)) == ones(Int,1,1)
  UndefVarError: .= not defined
   in matrix_repr(::Nemo.SkewDiagram) at
/home/wbhart/.julia/v0.5/Nemo/src/generic/YoungTabs.jl:360
   in test_skewdiags() at
/home/wbhart/.julia/v0.5/Nemo/test/../test/generic/YoungTabs-test.jl:90
   in test_ytabs at
/home/wbhart/.julia/v0.5/Nemo/test/../test/generic/YoungTabs-test.jl:169
[inlined]
   in test_groups() at
/home/wbhart/.julia/v0.5/Nemo/test/../test/Groups-test.jl:5
   in test_all() at
/home/wbhart/.julia/v0.5/Nemo/test/../test/Nemo-test.jl:7
   in include_from_node1(::String) at ./loading.jl:426
   in process_options(::Base.JLOptions) at ./client.jl:262
   in _start() at ./client.jl:318
ERROR: LoadError: There was an error during testing
 in record(::Base.Test.FallbackTestSet, ::Base.Test.Error) at ./test.jl:397
 in do_test(::Base.Test.Threw, ::Expr) at ./test.jl:281
 in test_skewdiags() at
/home/wbhart/.julia/v0.5/Nemo/test/../test/generic/YoungTabs-test.jl:90
 in test_ytabs at
/home/wbhart/.julia/v0.5/Nemo/test/../test/generic/YoungTabs-test.jl:169
[inlined]
 in test_groups() at
/home/wbhart/.julia/v0.5/Nemo/test/../test/Groups-test.jl:5
 in test_all() at /home/wbhart/.julia/v0.5/Nemo/test/../test/Nemo-test.jl:7
 in include_from_node1(::String) at ./loading.jl:426
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /home/wbhart/.julia/v0.5/Nemo/test/runtests.jl, in expression
starting on line 5

**Comment by thofma** [comment]
Which julia version?

**Comment by wbhart** [comment]
0.5

**Comment by kalmarek** [comment]
hmm... No idea how `.=` could not be defined in Your julia. Could You try running:
```julia
using Nemo
ep = Partition(Int[])
onep = Partition(Int[1])
xi = onep/ep;
@assert typeof(xi) == Nemo.SkewDiagram
ydiag = zeros(Int,length(xi.lam), maximum(xi.lam))
for i in 1:length(xi.mu)
   ydiag[i, xi.mu[i]+1:xi.lam[i]] .= 1
end
@assert ydiag == zeros(Int, 1,1)
for i in length(xi.mu)+1:length(xi.lam)
   ydiag[i,1:xi.lam[i]] .= 1
end
@assert ydiag == ones(Int, 1,1)
```
This is exactly what the test performs

**Comment by wbhart** [comment]
Yes, I get:

julia> for i in length(xi.mu)+1:length(xi.lam)
          ydiag[i,1:xi.lam[i]] .= 1
       end
ERROR: UndefVarError: .= not defined
 in macro expansion; at ./REPL[9]:2 [inlined]
 in anonymous at ./<missing>:?
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46

julia> @assert ydiag == ones(Int, 1,1)
ERROR: AssertionError: ydiag == ones(Int,1,1)
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46

Perhaps my Julia is just too old. I don't think it is a release 0.5, but a
dev version. Probably ignore this. My Julia is probably too old.

**Pull request merged**